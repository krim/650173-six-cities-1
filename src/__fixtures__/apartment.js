export default {
  city: {
    name: `Amsterdam`,
    location: {
      latitude: 51.225402,
      longitude: 6.776314,
      zoom: 13
    }
  },
  previewImage: `https://es31-server.appspot.com/six-cities/static/hotel/1.jpg`,
  images: [
    `https://es31-server.appspot.com/six-cities/static/hotel/18.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/4.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/12.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/17.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/15.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/2.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/8.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/5.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/6.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/1.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/7.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/19.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/14.jpg`
  ],
  title: `Waterfront with extraordinary view`,
  isFavorite: false,
  isPremium: true,
  rating: 2.7,
  type: `hotel`,
  bedrooms: 2,
  maxAdults: 8,
  price: 393,
  goods: [
    `Breakfast`,
    `Washer`,
    `Laptop friendly workspace`
  ],
  host: {
    id: 25,
    name: `Angelina`,
    isPro: true,
    avatarUrl: `img/avatar-angelina.jpg`
  },
  description: `I am happy to welcome you to my apartment in the city center! Three words: location, cosy and chic!`,
  location: {
    latitude: 51.214402,
    longitude: 6.764314000000001,
    zoom: 16
  },
  id: 1
};
