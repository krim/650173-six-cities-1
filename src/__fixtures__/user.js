export default {
  id: 13,
  isPro: false,
  name: `Zak`,
  avatarUrl: `https://es31-server.appspot.com/six-cities/static/avatar/4.jpg`
};
