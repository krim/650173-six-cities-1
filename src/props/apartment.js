import PropTypes from 'prop-types';
import {range} from 'react-range-proptypes';

export default PropTypes.shape({
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  previewImage: PropTypes.string.isRequired,
  rating: range(0, 100).isRequired,
  price: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  bedrooms: PropTypes.number.isRequired,
  maxAdults: PropTypes.number.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  isPremium: PropTypes.bool.isRequired,
  location: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
    zoom: PropTypes.number.isRequired
  }),
  host: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    isPro: PropTypes.bool.isRequired,
    avatarUrl: PropTypes.string.isRequired
  })
});
