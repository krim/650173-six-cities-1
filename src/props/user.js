import PropTypes from 'prop-types';

export default PropTypes.shape({
  id: PropTypes.number,
  name: PropTypes.string,
  email: PropTypes.string,
  avatarUrl: PropTypes.string,
  isPro: PropTypes.bool
});
