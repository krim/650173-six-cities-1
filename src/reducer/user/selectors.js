// import {createSelector} from 'reselect';
import NameSpace from '../name-spaces';

const NAME_SPACE = NameSpace.USER;

export const getAuthorizationRequired = (state) => state[NAME_SPACE].isAuthorizationRequired;
export const getUser = (state) => state[NAME_SPACE].user;
