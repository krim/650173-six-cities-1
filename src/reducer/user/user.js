import camelcaseKeys from 'camelcase-keys';

const initialState = {
  user: {}
};

const ActionType = {
  AUTHORIZATION: `AUTHORIZATION`,
  CHECK_AUTHORIZATION: `CHECK_AUTHORIZATION`
};

const ActionCreator = {
  authorize: (data) => ({
    type: ActionType.AUTHORIZATION,
    payload: data
  })
};

const Operation = {
  authorize: (data, ownProps) => (dispatch, _getState, api) => {
    return api.post(`/login`, data)
      .then((response) => {
        const userData = camelcaseKeys(response.data);
        dispatch(ActionCreator.authorize(userData));
        ownProps.history.push(`/`);
      }).
      catch((_error) => {});
  },
  checkAuthorization: () => (dispatch, _getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        const userData = camelcaseKeys(response.data);
        dispatch(ActionCreator.authorize(userData));
      }).
      catch((_error) => {});
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.AUTHORIZATION: return {...state, user: action.payload};
  }

  return state;
};

export {
  ActionCreator,
  ActionType,
  Operation,
  reducer
};
