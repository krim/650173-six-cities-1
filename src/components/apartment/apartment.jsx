import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

import {apartmentProps} from '../../props';
import getRating from '../../libs/get-rating';

class Apartment extends PureComponent {
  render() {
    const {
      apartment,
      onClick,
      onMouseOver,
      onMouseOut
    } = this.props;

    return (
      <article
        onMouseOver={onMouseOver}
        onMouseOut={onMouseOut}
        className="cities__place-card place-card"
      >
        { apartment.isPremium && <div className="place-card__mark"><span>Premium</span></div> }
        <div className="cities__image-wrapper place-card__image-wrapper">
          <a href="#">
            <img
              className="place-card__image" src={apartment.previewImage} width="260" height="200" alt="Place image"
            />
          </a>
        </div>
        <div className="place-card__info">
          <div className="place-card__price-wrapper">
            <div className="place-card__price">
              <b className="place-card__price-value">&euro;{apartment.price}</b>
              <span className="place-card__price-text">&#47;&nbsp;night</span>
            </div>
            <button className="place-card__bookmark-button button" type="button">
              <svg className="place-card__bookmark-icon" width="18" height="19">
                <use xlinkHref="#icon-bookmark"></use>
              </svg>
              <span className="visually-hidden">To bookmarks</span>
            </button>
          </div>
          <div className="place-card__rating rating">
            <div className="place-card__stars rating__stars">
              <span style={{width: getRating(apartment.rating)}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <h2 className="place-card__name">
            <a href="#" onClick={onClick}>{ apartment.title }</a>
          </h2>
          <p className="place-card__type">{apartment.type}</p>
        </div>
      </article>
    );
  }
}

Apartment.propTypes = {
  apartment: apartmentProps,
  onClick: PropTypes.func.isRequired,
  onMouseOver: PropTypes.func.isRequired,
  onMouseOut: PropTypes.func.isRequired
};

export default Apartment;

