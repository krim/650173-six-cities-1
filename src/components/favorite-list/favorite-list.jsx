import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {Operation} from '../../reducer/favorite/favorite';
import {getFavorites} from '../../reducer/favorite/selectors';
import {apartmentProps} from '../../props';
import Apartment from '../apartment/apartment.jsx';
import FavoriteEmptyList from '../favorite-empty-list/favorite-empty-list.jsx';

class FavoriteList extends PureComponent {
  componentDidMount() {
    this.props.loadFavorites();
  }

  render() {
    const {favorites} = this.props;

    if (favorites.length === 0) {
      return <FavoriteEmptyList />;
    }

    return (
      <main className="page__main page__main--favorites">
        <div className="page__favorites-container container">
          <section className="favorites">
            <h1 className="favorites__title">Saved listing</h1>
            <ul className="favorites__list">
              {
                favorites.map((favorite) => {
                  return (
                    <li key={favorite.city} className="favorites__locations-items">
                      <div className="favorites__locations locations locations--current">
                        <div className="locations__item">
                          <a className="locations__item-link" href="#">
                            <span>{favorite.city}</span>
                          </a>
                        </div>
                      </div>
                      <div className="favorites__places">
                        {this._renderApartments(favorite.apartments)}
                      </div>
                    </li>
                  );
                })
              }
            </ul>
          </section>
        </div>
      </main>
    );
  }

  _renderApartments(apartments) {
    return apartments.map((apartment) => {
      return (
        // TODO: передавать класс favorites
        <Apartment
          key={`apartment-${apartment.id}`}
          apartment={apartment}
          onClick={() => {}}
          setActive={() => {}}
          disableActive={() => {}}
        />
      );
    });
  }
}

FavoriteList.propTypes = {
  favorites: PropTypes.arrayOf(
      PropTypes.shape({
        city: PropTypes.string.isRequired,
        apartments: PropTypes.arrayOf(apartmentProps).isRequired
      })
  ),
  loadFavorites: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
  return {
    favorites: getFavorites(state)
  };
};

const mapDispatchToProps = (dispatch) => ({
  loadFavorites: () => dispatch(Operation.loadFavorites())
});

export {FavoriteList};
export default connect(mapStateToProps, mapDispatchToProps)(FavoriteList);
