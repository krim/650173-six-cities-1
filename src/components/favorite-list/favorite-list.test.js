import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter as Router} from 'react-router-dom';

import apartment from '../../__fixtures__/apartment';
import user from '../../__fixtures__/user';
import {FavoriteList} from './favorite-list';

jest.mock(`../../hocs/with-favorite/with-favorite`, () => (component) => component);

const favorites = [{city: `Amsterdam`, apartments: [apartment]}];

describe(`FavoriteList`, () => {
  it(`renders component correctly`, () => {
    const tree = renderer.
      create(
          <Router>
            <FavoriteList
              user={user}
              favorites={favorites}
              loadFavorites={jest.fn()}
            />
          </Router>
      ).
      toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe(`when there are no favorite apartments`, () => {
    it(`renders component correctly`, () => {
      const tree = renderer.
        create(
            <Router>
              <FavoriteList
                user={user}
                favorites={[]}
                loadFavorites={jest.fn()}
              />
            </Router>
        ).
        toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
