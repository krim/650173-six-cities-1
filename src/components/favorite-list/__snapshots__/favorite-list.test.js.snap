// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteList renders component correctly 1`] = `
<main
  className="page__main page__main--favorites"
>
  <div
    className="page__favorites-container container"
  >
    <section
      className="favorites"
    >
      <h1
        className="favorites__title"
      >
        Saved listing
      </h1>
      <ul
        className="favorites__list"
      >
        <li
          className="favorites__locations-items"
        >
          <div
            className="favorites__locations locations locations--current"
          >
            <div
              className="locations__item"
            >
              <a
                className="locations__item-link"
                href="#"
              >
                <span>
                  Amsterdam
                </span>
              </a>
            </div>
          </div>
          <div
            className="favorites__places"
          >
            <article
              className="cities__place-card place-card"
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="https://es31-server.appspot.com/six-cities/static/hotel/1.jpg"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      393
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button"
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "60%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="#"
                    onClick={[Function]}
                  >
                    Waterfront with extraordinary view
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  hotel
                </p>
              </div>
            </article>
          </div>
        </li>
      </ul>
    </section>
  </div>
</main>
`;

exports[`FavoriteList when there are no favorites apartments renders component correctly 1`] = `
<main
  className="page__main page__main--favorites page__main--favorites-empty"
>
  <div
    className="page__favorites-container container"
  >
    <section
      className="favorites favorites--empty"
    >
      <h1
        className="visually-hidden"
      >
        Favorites (empty)
      </h1>
      <div
        className="favorites__status-wrapper"
      >
        <b
          className="favorites__status"
        >
          Nothing yet saved.
        </b>
        <p
          className="favorites__status-description"
        >
          Save properties to narrow down search or plan yor future trips.
        </p>
      </div>
    </section>
  </div>
</main>
`;
